src/EditorFeatures/Core/InheritanceMargin/AbstractInheritanceMarginService.cs

```diff
@@ -0,0 +1,159 @@
+            var mappingService = document.Project.Solution.Workspace.Services.GetRequiredService<ISymbolMappingService>();
+            using var _ = ArrayBuilder<InheritanceMemberItem>.GetInstance(out var builder);
+
+            // Iterate all the members symbol to find each their inheritance chain information
+            foreach (var memberDeclarationNode in allDeclarationNodes)
+            {
+                var member = semanticModel.GetDeclaredSymbol(memberDeclarationNode, cancellationToken);
+                if (member != null && !member.IsStatic && !member.IsErrorType())
+                {
+                    var mappingResult = await mappingService.MapSymbolAsync(document, member, cancellationToken).ConfigureAwait(false);
+                    if (mappingResult != null)
+                    {
+                        if (mappingResult.Symbol is INamedTypeSymbol namedTypeSymbol)
+                        {
+                            // Find its baseTypes and subTypes.
+                            await AddInheritanceMemberItemsForNamedTypeAsync(
+                                mappingResult.Project.Solution,
+                                sourceText,
+                                memberDeclarationNode,
+                                namedTypeSymbol, builder, cancellationToken).ConfigureAwait(false);
+                        }
+
+                        if (mappingResult.Symbol is IEventSymbol or IPropertySymbol || mappingResult.Symbol.IsOrdinaryMethod())
+                        {
+                            // Find the implementing/implemented/overriden/overriding members
+                            await AddInheritanceMemberItemsForTypeMembersAsync(
+                                mappingResult.Project.Solution,
+                                sourceText,
+                                memberDeclarationNode,
+                                mappingResult.Symbol,
+                                builder,
+                                cancellationToken).ConfigureAwait(false);
+                        }
+                    }
+                }
+            }
+
+            return builder.ToImmutable();
+        }
+
+        private static async Task AddInheritanceMemberItemsForNamedTypeAsync(
+            Solution solution,
+            SourceText sourceText,
+            SyntaxNode declarationNode,
+            INamedTypeSymbol memberSymbol,
+            ArrayBuilder<InheritanceMemberItem> builder,
+            CancellationToken cancellationToken)
+        {
+            // Get all base types
+            var allBaseTypes = BaseTypeFinder.FindBaseTypesAndInterfaces(memberSymbol);
+
+            // Filter out System.Object. (otherwise margin would be shown for all classes)
+            var baseTypes = allBaseTypes
+                .WhereAsArray(type =>
+                    !(type is ITypeSymbol { SpecialType: SpecialType.System_Object }));
```

Comment:
❔ What about structs inheriting from `ValueType`, and enums inheriting from `Enum`? Does the UI still show?

Reply:
```suggestion
            // Filter out System.Object. (otherwise margin would be shown for all classes)
            var baseTypes = allBaseTypes
                .WhereAsArray(type =>
                    type is not { SpecialType: SpecialType.System_Object }));
```

Reply:
or, more simply: `type.SpecialType != System_Object`

Reply:
i'm not a fan of wordy pattern matching for the types of checks that are already easy in hte language.

Reply:
: ) To Sam: I can add unit test to cover the case you are worrying about.
 To Cyrus: sure let me refactor this part